x/stake/cross_stake/cross_stake.go

```diff
@@ -0,0 +1,340 @@
+package cross_stake
+
+import (
+	"strconv"
+
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	"github.com/cosmos/cosmos-sdk/bsc"
+	"github.com/cosmos/cosmos-sdk/bsc/rlp"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/stake/types"
+)
+
+type CrossStakeApp struct {
+	stakeKeeper Keeper
+}
+
+func NewCrossStakeApp(stakeKeeper Keeper) *CrossStakeApp {
+	return &CrossStakeApp{
+		stakeKeeper: stakeKeeper,
+	}
+}
+
+func (app *CrossStakeApp) ExecuteAckPackage(ctx sdk.Context, payload []byte) sdk.ExecuteResult {
+	if len(payload) == 0 {
+		app.stakeKeeper.Logger(ctx).Info("receive cross stake ack package")
+		return sdk.ExecuteResult{}
+	}
+
+	eventCode, pack, err := DeserializeCrossStakeSynPackage(payload)
+	if err != nil {
+		app.stakeKeeper.Logger(ctx).Error("unmarshal cross stake ack package error", "err", err.Error(), "package", string(payload))
+		panic("unmarshal cross stake ack package error")
+	}
+
+	var result sdk.ExecuteResult
+	switch eventCode {
+	case types.CrossStakeTypeDistributeReward:
+		result, err = app.handleDistributeRewardRefund(ctx, pack.(types.CrossStakeDistributeRewardSynPackage))
+	case types.CrossStakeTypeDistributeUndelegated:
+		result, err = app.handleDistributeUndelegatedRefund(ctx, pack.(types.CrossStakeDistributeUndelegatedSynPackage))
+	}
+	if err != nil {
+		panic(err)
+	}
+
+	return result
+}
+
+func (app *CrossStakeApp) ExecuteFailAckPackage(ctx sdk.Context, payload []byte) sdk.ExecuteResult {
+	app.stakeKeeper.Logger(ctx).Error("received cross stake fail ack package ")
+	return sdk.ExecuteResult{}
+}
+
+func (app *CrossStakeApp) ExecuteSynPackage(ctx sdk.Context, payload []byte, relayFee int64) sdk.ExecuteResult {
+	app.stakeKeeper.Logger(ctx).Info("receive cross stake syn package")
+	eventCode, pack, err := DeserializeCrossStakeSynPackage(payload)
+	if err != nil {
+		app.stakeKeeper.Logger(ctx).Error("unmarshal cross stake sync claim error", "err", err.Error(), "claim", string(payload))
+		panic("unmarshal cross stake claim error")
+	}
+
+	var result sdk.ExecuteResult
+	switch eventCode {
+	case types.CrossStakeTypeDelegate:
+		result, err = app.handleDelegate(ctx, pack.(types.CrossStakeDelegateSynPackage), relayFee)
+	case types.CrossStakeTypeUndelegate:
+		result, err = app.handleUndelegate(ctx, pack.(types.CrossStakeUndelegateSynPackage), relayFee)
+	case types.CrossStakeTypeRedelegate:
```

Comment:
Please throw exception if the eventCode is unknown.

Reply:
Done